File: 백준/Silver/1991. 트리 순회/트리 순회.py
import sys


def preorder(n):
    if n:
        print(chr(n + 64), end='')
        preorder(ch1[n])
        preorder(ch2[n])


def inorder(n):
    if n:
        inorder(ch1[n])
        print(chr(n + 64), end='')
        inorder(ch2[n])


def postorder(n):
    if n:
        postorder(ch1[n])
        postorder(ch2[n])
        print(chr(n + 64), end='')


N = int(sys.stdin.readline())
ch1 = [0] * (N+1)
ch2 = [0] * (N+1)
for _ in range(1, N+1):
    v, cl, cr = sys.stdin.readline().split()
    if cl == '.':
        ch1[ord(v) - 64] = 0
    else:
        ch1[ord(v) - 64] = ord(cl) - 64
    if cr == '.':
        ch2[ord(v) - 64] = 0
    else:
        ch2[ord(v) - 64] = ord(cr) - 64

preorder(1)
print()
inorder(1)
print()
postorder(1)
print()

File: 백준/Silver/1991. 트리 순회/트리 순회.java
import java.util.*;
import java.io.*;


public class Main {
	
	static void preOrder(char node) {
		if (node != '.') {
			System.out.print(node);
			preOrder(tree.get(node)[0]);
			preOrder(tree.get(node)[1]);
		}
	}

	static void inOrder(char node) {
		if (node != '.') {
			inOrder(tree.get(node)[0]);
			System.out.print(node);
			inOrder(tree.get(node)[1]);
		}
	}

	static void postOrder(char node) {
		if (node != '.') {
			postOrder(tree.get(node)[0]);
			postOrder(tree.get(node)[1]);
			System.out.print(node);
		}
	}
	
	static Map<Character, char[]> tree;
	
    public static void main(String[] args) throws IOException {
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	
    	int N = Integer.parseInt(br.readLine());
    	tree = new HashMap<>();
    	for (int i = 0; i < N; i++) {
    		StringTokenizer st = new StringTokenizer(br.readLine());
    		char root = st.nextToken().charAt(0);
    		char left = st.nextToken().charAt(0);
    		char right = st.nextToken().charAt(0);
    		
    		tree.put(root, new char[] {left, right});
    	}
    	
    	preOrder('A');
    	System.out.println();
    	inOrder('A');
    	System.out.println();
    	postOrder('A');
    	
    	br.close();
    }
}
